### PR TITLE
layernorm & groupnorm bwd gamma beta

### PR DESCRIPTION
**Pull request by rocking5566**
ckprofiler, test and client example

### CODE REVIEW

File: client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#include <iomanip>
+#include <vector>
+#include <iostream>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
```
Comment (qianfengz) [inline client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp]: Including `tensor_layout.hpp` seems not needed in this file

File: client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+#include "ck/tensor_operation/gpu/device/device_normalization_bwd_gamma_beta.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
```
Comment (qianfengz) [inline client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp]: Including `element_wise_operation.hp` seems not needed in this file

File: client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#include <iomanip>
+#include <vector>
+#include <iostream>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
```
Comment (qianfengz) [inline client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp]: Including of `tensor_layout.hpp` seems not needed in this file

File: client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+#include "ck/tensor_operation/gpu/device/device_normalization_bwd_gamma_beta.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
```
Comment (qianfengz) [inline client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp]: Including `element_wise_operation.hpp` seems not needed in this file

File: library/include/ck/library/tensor_operation_instance/gpu/groupnorm_bwd_gamma_beta.hpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#pragma once
+
+#include <vector>
+#include <memory>
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
```
Comment (qianfengz) [inline library/include/ck/library/tensor_operation_instance/gpu/groupnorm_bwd_gamma_beta.hpp:9]: Including `tensor_layout.hpp` seems not needed in this file

File: library/include/ck/library/tensor_operation_instance/gpu/layernorm_bwd_gamma_beta.hpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
+
+#pragma once
+
+#include <vector>
+#include <memory>
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
```
Comment (qianfengz) [inline library/include/ck/library/tensor_operation_instance/gpu/layernorm_bwd_gamma_beta.hpp]: Including `tensor_layout.hpp` seems not needed in this file

File: profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        mean.GenerateTensorValue(GeneratorTensor_2<MeanInvStdDataType>{-5, 5});
+        inv_std.GenerateTensorValue(GeneratorTensor_2<MeanInvStdDataType>{-5, 5});
+        dgamma.GenerateTensorValue(GeneratorTensor_2<DGammaDataType>{-5, 5});
+        dbeta.GenerateTensorValue(GeneratorTensor_2<DBetaDataType>{-5, 5});
+        break;
+    default:
+        dy.GenerateTensorValue(GeneratorTensor_3<DYDataType>{0, 1});
+        x.GenerateTensorValue(GeneratorTensor_3<XDataType>{0, 1});
+        mean.GenerateTensorValue(GeneratorTensor_3<MeanInvStdDataType>{-0.5, 0.5});
+        inv_std.GenerateTensorValue(GeneratorTensor_3<MeanInvStdDataType>{-0.5, 0.5});
```
Comment (qianfengz) [inline profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp]: Actually,  `inv_std` could never be negative values

File: profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        mean.GenerateTensorValue(GeneratorTensor_1<MeanInvStdDataType>{});
+        inv_std.GenerateTensorValue(GeneratorTensor_1<MeanInvStdDataType>{});
+        dgamma.GenerateTensorValue(GeneratorTensor_1<DGammaDataType>{});
+        dbeta.GenerateTensorValue(GeneratorTensor_1<DBetaDataType>{});
+        break;
+    case 1:
+        dy.GenerateTensorValue(GeneratorTensor_2<DYDataType>{-5, 5});
+        x.GenerateTensorValue(GeneratorTensor_2<XDataType>{-5, 5});
+        mean.GenerateTensorValue(GeneratorTensor_2<MeanInvStdDataType>{-5, 5});
+        inv_std.GenerateTensorValue(GeneratorTensor_2<MeanInvStdDataType>{-5, 5});
```
Comment (qianfengz) [inline profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp]: Actually `inv_std` can never be negative values

File: profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        auto invoker_ptr = inst_ptr->MakeInvokerPointer();
+
+        float avg_time = invoker_ptr->Run(argument_ptr.get(), StreamConfig{nullptr, time_kernel});
+
+        std::size_t num_bytes = dy.mDesc.GetElementSize() * sizeof(DYDataType) +
+                                x.mDesc.GetElementSize() * sizeof(XDataType) +
+                                mean.mDesc.GetElementSize() * sizeof(MeanInvStdDataType) +
+                                inv_std.mDesc.GetElementSize() * sizeof(MeanInvStdDataType) +
+                                dgamma.mDesc.GetElementSize() * sizeof(DGammaDataType) +
+                                dbeta.mDesc.GetElementSize() * sizeof(DBetaDataType);
```
Comment (qianfengz) [inline profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp]: `num_bytes` should be calculated outside the for loop

File: profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        auto invoker_ptr = inst_ptr->MakeInvokerPointer();
+
+        float avg_time = invoker_ptr->Run(argument_ptr.get(), StreamConfig{nullptr, time_kernel});
+
+        std::size_t num_bytes = dy.mDesc.GetElementSize() * sizeof(DYDataType) +
+                                x.mDesc.GetElementSize() * sizeof(XDataType) +
+                                mean.mDesc.GetElementSize() * sizeof(MeanInvStdDataType) +
+                                inv_std.mDesc.GetElementSize() * sizeof(MeanInvStdDataType) +
+                                dgamma.mDesc.GetElementSize() * sizeof(DGammaDataType) +
+                                dbeta.mDesc.GetElementSize() * sizeof(DBetaDataType);
```
Comment (qianfengz) [inline profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp]: `num_bytes` should be calculated outside the for loop

File: client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+            size_t workspace_sz = op_ptr->GetWorkSpaceSize(argument_ptr.get());
+            SimpleDeviceMem workspace(workspace_sz);
+            op_ptr->SetWorkSpacePointer(argument_ptr.get(), workspace.GetDeviceBuffer());
+
+            float ave_time = invoker_ptr->Run(argument_ptr.get(), StreamConfig{nullptr, true});
+
+            std::size_t num_byte = sizeof(DYDataType) * length + sizeof(XDataType) * length +
+                                   sizeof(GammaDataType) * G * C +
+                                   sizeof(MeanInvStdDataType) * N * G * 2 +
+                                   sizeof(DGammaDataType) * G * C + sizeof(DBetaDataType) * G * C;
```
Comment (qianfengz) [inline client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp]: Rename to `num_bytes` and move this calculation outside the loop

File: client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp (hunk 1 above)
Comment (rocking5566) [reply client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp]: 3b70e18b82dc179d9a93747f94cdb8374366b14e

File: client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp (hunk 2 above)
Comment (rocking5566) [reply client_example/05_layernorm/layernorm2d_bwd_gamma_beta.cpp]: 3b70e18b82dc179d9a93747f94cdb8374366b14e

File: client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp (hunk 3 above)
Comment (rocking5566) [reply client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp]: 3b70e18b82dc179d9a93747f94cdb8374366b14e

File: library/include/ck/library/tensor_operation_instance/gpu/layernorm_bwd_gamma_beta.hpp (hunk 6 above)
Comment (rocking5566) [reply library/include/ck/library/tensor_operation_instance/gpu/layernorm_bwd_gamma_beta.hpp]: 3b70e18b82dc179d9a93747f94cdb8374366b14e

File: library/include/ck/library/tensor_operation_instance/gpu/groupnorm_bwd_gamma_beta.hpp (hunk 5 above)
Comment (rocking5566) [reply library/include/ck/library/tensor_operation_instance/gpu/groupnorm_bwd_gamma_beta.hpp:9]: 3b70e18b82dc179d9a93747f94cdb8374366b14e

File: client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp (hunk 4 above)
Comment (rocking5566) [reply client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp]: 3b70e18b82dc179d9a93747f94cdb8374366b14e

File: profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp (hunk 7 above)
Comment (rocking5566) [reply profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp]: 0ceee37dfa98d47797fd77642f287ce1343f4cd2

File: profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp (hunk 8 above)
Comment (rocking5566) [reply profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp]: 0ceee37dfa98d47797fd77642f287ce1343f4cd2

File: profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp (hunk 9 above)
Comment (rocking5566) [reply profiler/include/profiler/profile_layernorm_bwd_gamma_beta_impl.hpp]: 12dbc7ce3e1daf1dbb8330ec4b29d341e19e096c

File: profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp (hunk 10 above)
Comment (rocking5566) [reply profiler/include/profiler/profile_groupnorm_bwd_gamma_beta_impl.hpp]: 12dbc7ce3e1daf1dbb8330ec4b29d341e19e096c

File: client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp (hunk 11 above)
Comment (rocking5566) [reply client_example/18_groupnorm/groupnorm_bwd_gamma_beta.cpp]: 12dbc7ce3e1daf1dbb8330ec4b29d341e19e096c

**Review by qianfengz — APPROVED** [review]

**Review by carlushuang — APPROVED** [review]
LGTM

**Pull request merged**